Created a node package for functions

evgeny-w opened the pull request:
Created a node package that exposes functions
>getProcList
>killProcByPID

evgeny-w committed:
Created a node package for functions

hello1101n commented on node-task-pkg/index.js:
```diff
@@ -0,0 +1,51 @@
+
+exports.getProcList = async function (username) {
+    let result = {stdout:"",stderr:""};
+    try{
```
I recommend you install some linter. Like, eslint and choose style guide like Airbnb for example.

hello1101n commented on node-task-pkg/index.js:
```diff
@@ -0,0 +1,51 @@
+exports.getProcList = async function (username) {
+    let result = {stdout:"",stderr:""};
+    try{
+        if (plat == "win32"){
```
I think we don't have win32 and others (if and else). We have 3 platforms. Maybe switch will be better? And default case: error because we don't work with this OS

evgeny-w replied:
Sorry I am confused what you mean, what platforms will we be using? When I request the platform version it is returned like this:
- Windows : win32
- Linux : linux
- Mac OS: darwin

Do we need to support other platforms except for these?

hello1101n replied:
A good practice is to handle unsupported os. Write error instead of doing nothing.

`os.platform() `can return 'aix', 'darwin', 'freebsd','linux', 'openbsd', 'sunos', and 'win32'
https://nodejs.org/api/os.html#osplatform

hello1101n replied:
Now we only need support for win32, linux, darwin

hello1101n commented on node-task-pkg/index.js:
```diff
@@ -0,0 +1,51 @@
+const os = require("os");
+const plat = os.platform();
+
+exports.getProcList = async function (username) {
```
After you will install the linter, it will show you that username - unused var.

hello1101n commented on node-task-pkg/index.js:
```diff
@@ -0,0 +1,51 @@
+const plat = os.platform();
+
+exports.getProcList = async function (username) {
+    let result = {stdout:"",stderr:""};
```
I think we don't want to know how we get out and err outside. Maybe rename to `processes` and `error`?

hello1101n commented on node-task-pkg/index.js:
```diff
@@ -0,0 +1,51 @@
+exports.killProcByPID = async function (pid) {
+    let result = {stdout:"",stderr:""};
+    try{
+        if (plat == "win32"){
```
Also comment about change if to switch case, about change stdout stderr names and add linter.

hello1101n replied:
== - not the best thing in javascript. If you don't have to use it. I recommend in basic case use === instead

hello1101n commented on node-task-pkg/index.js:
```diff
@@ -0,0 +1,51 @@
+    
+    return result;
+};
+
```
Better to add typescript in future, but now you can add comment before functions. Something like https://jsdoc.app/tags-example.html

hello1101n commented on node-task-pkg/index.js:
```diff
@@ -0,0 +1,51 @@
+    try{
+        if (plat == "win32"){
+            result = await execProm("C:/Windows/System32/tasklist.exe /FO CSV")
+            result.stdout = result.stdout.trim().split("\r\n")
```
I think we can use array method map here instead of for loop. It will be better to read

hello1101n commented on node-task-pkg/index.js:
```diff
@@ -0,0 +1,51 @@
+                result = await execProm("ps -e -o pid,command | awk '{printf \"%s,\",$1;$1=\"\";print substr($0,2)}'")
+                result.stdout = result.stdout.trim().split("\n")
+            } else if (plat == "darwin"){
+                result = await execProm("ps -ec -o pid,command | awk '{printf \"%s,\",$1;$1=\"\";print substr($0,2)}'")
```
Same string in 19 line. I think we can create var for it.

evgeny-w replied:
The strings are slightly different, line 19 invokes ps -e while line 22 invokes ps -ec. From my testing, I found that -ec is better on Mac but does not work properly on Linux.

hello1101n replied:
Ok

hello1101n commented on node-task-pkg/index.js:
```diff
@@ -0,0 +1,51 @@
+    }
+    
+    
+    return result;
```
Also, we can avoid mutation "result" var. We can return when we calculate the process list or error. It will be easier to read.

evgeny-w replied:
I am not sure what you mean, do you mean that it would be better to return immediately after getting the list of processes (inside of the switch statement) or to not use a structure to return result/error pair?

hello1101n replied:
I mean instead of this:

```js
let result = {result: '', error: ''}

case ***:
  result.error = ''
  result.result = ***
  break
case ***:
  result.error = ''
  result.result = ***
   break
```

Do this:
```
case ***:
  result = ***
  return {result, error: ''}
case ***:
  result = ***
  return {result, error: ''}
`
```

It is more straightforwardly. Because you don't need to see all function to see where you want to mutate result object.

hello1101n commented on node-task-pkg/index.js:
```diff
@@ -0,0 +1,51 @@
+        result.stdout="ERROR";
+        result.stderr=ex;
+    }
+    return result;
```
Also, we can avoid mutation "result" var. We can return when we calculate the process list or error. It will be easier to read. x2

hello1101n commented on node-task-pkg/package.json:
```diff
@@ -0,0 +1,24 @@
+{
+   "name":"node-task-mgr",
+   "version":"1.0.1",
+   "description":"A cross-platform node module that provides tools to get running process info",
```
Get and kill!)

hello1101n commented on node-task-pkg/package.json:
```diff
@@ -0,0 +1,24 @@
+   "description":"A cross-platform node module that provides tools to get running process info",
+   "main":"index.js",
+   "scripts":{
+      "test":"echo \"Error: no test specified\" && exit 1"
```
Do you need this?

hello1101n commented:
After this review let's merge it and i will implement it in sandbox app to send to QA.

hello1101n commented:
Also let's add good commit messages to master. Please use squash to create better commit messages. Also i recommend style like in [linux](https://github.com/torvalds/linux/commits/master) or [vue](https://github.com/vuejs/vue/commits/main).

Something like:
module: folder/file: what did you do. 

Or: 
module: what did you do

evgeny-w committed:
node-task-pkg: index.js:
>Formatted code with ESLint Airbnb style
>Added JSDOC documentation
>Changed return value of functions to be more readable
>>getProcList -> output.processes & output.error
>>killProcByPID -> output.result & output.error
>Added a case handler for unsupported OS
>Replaced checks for platform with switch statements (instead of if)
>Implemented array map function when processing process list

node-task-pkg: package.json:
>Removed tests
>Adjusted description

Signed-off-by: Evgeny <evgeny.vereshchagin@alemira.com>

evgeny-w commented:
Hi Nikita,
I implemented most of your suggestions, thank you. I had some issues with understanding some of them, and I commented underneath them, please take a look. I also have another question, what do you mean by "using" squash to create better commit messages, is squash a tool for making commit messages? I am confused by Google results about this. Sorry for the commit messages on master, I was trying to move the local project into github, so there were no meaningful messages I could add (I was not working with a source control before).

I will now push ESLinted main branch, so the commit message for all the files should become idential if that is what you mean.

hello1101n commented:
> I also have another question, what do you mean by "using" squash to create better commit messages, is squash a tool for making commit messages

What do you think about meeting with me today at 11 gmt+3 (am)?  I think i will be easier to do with pair programming

hello1101n commented on node-task-pkg/index.js:
```diff
@@ -0,0 +1,154 @@
+  const output = new ProcessOutputFormat();
+  let result = { stdout: '', stderr: '' };
+  try {
+    let i;
```
unused var?

hello1101n commented on node-task-pkg/index.js:
```diff
@@ -0,0 +1,154 @@
+        output.processes = result.stdout.trim().split('\n').map(x => x.trim().split(','));
+        break;
+      default:
+        throw new OperatingSystemNotSupportedException();
```
Does it work correctly?

hello1101n commented on node-task-pkg/index.js:
```diff
@@ -0,0 +1,154 @@
+        result = await execProm(`kill -9 ${pid}`);
+        break;
+      default:
+        throw new OperatingSystemNotSupportedException();
```
Does it work correctly?x2

hello1101n commented on node-task-pkg/index.js:
```diff
@@ -0,0 +1,154 @@
+  return output;
+};
+
+exports.getProcList().then((result) => {
```
Why do you need this?

hello1101n replied:
If this is only for test purpose - you can delete it

hello1101n commented on node-task-pkg/index.js:
```diff
@@ -0,0 +1,154 @@
+/**
+ * Return value for getProcList function
+ * @constructor
+ * @returns {String} processes
```
processes: [
 * //       ['PID','Image Name'],
 * //       ['0','System Idle Process'],
 * //       ... more items
 * //     ],
 
 Looks like processes not just String

hello1101n commented on node-task-pkg/index.js:
```diff
@@ -0,0 +1,154 @@
+ * getProcList();
+ * @example
+ * // returns {
+ * //     processes: 'ERROR',
```
Maybe empty list is better?

hello1101n commented on node-task-pkg/index.js:
```diff
@@ -0,0 +1,154 @@
+ * killProcByPID(0);
+ * @example
+ * // returns {
+ * //     processes: 'ERROR',
```
processes? This is killProcByPID function

hello1101n commented:
Also i want meeting to do good with "This branch has conflicts that must be resolved". And show you how.

evgeny-w committed:
node-task-pkg: index.js
>Fixed incorrect JsDoc identifiers
>Removed unused variables and "result" variable mutation
>Fixed vulnerabilities in Windows and Linux process list commands
>Adjusted return values of functions at errors
>Fixed vulnerability of passing non-number PID to be executed with command line
>Commented out testing code


Signed-off-by: Evgeny <evgeny.vereshchagin@alemira.com>

evgeny-w committed:
Removed unnececary comments

evgeny-w committed:
node-task-pkg: index.js
>Fixed mutation of result variable
>Adjusted testing of PID to be a number to fix empty strings being a number issue


Signed-off-by: Evgeny <evgeny.vereshchagin@alemira.com>

evgeny-w committed:
node-task-pkg: index.js: Added kernel-based support for linux systems
>Made output consistent for Windows and Linux, TODO: Mac


Signed-off-by: Evgeny <evgeny.vereshchagin@alemira.com>

evgeny-w committed:
node-task-pkg: index.js:
>Improved input verification for killProcByPID
>Adjusted list of process command for Mac
>Removed redundancy


Signed-off-by: Evgeny <evgeny.vereshchagin@alemira.com>

evgeny-w committed:
node-task-pkg: index.js:
>Improved OS error readability
>Improved function variable mutation readibility issues


Signed-off-by: Evgeny <evgeny.vereshchagin@alemira.com>
